qiskit/visualization/pulse_v2/events.py

```diff
@@ -145,48 +151,117 @@ def is_empty(self):
         sorted_frame_changes = sorted(self._frames.items(), key=lambda x: x[0], reverse=True)
         sorted_waveforms = sorted(self._waveforms.items(), key=lambda x: x[0])
 
         # bind phase and frequency with instruction
-        phase = self.init_phase
-        frequency = self.init_frequency
+        phase = self._init_phase
+        frequency = self._init_frequency
         for t0, inst in sorted_waveforms:
             while len(sorted_frame_changes) > 0 and sorted_frame_changes[-1][0] <= t0:
                 _, frame_changes = sorted_frame_changes.pop()
-                for frame_change in frame_changes:
-                    if isinstance(frame_change, pulse.instructions.SetFrequency):
-                        frequency = frame_change.frequency
-                    elif isinstance(frame_change, pulse.instructions.ShiftFrequency):
-                        frequency += frame_change.frequency
-                    elif isinstance(frame_change, pulse.instructions.SetPhase):
-                        phase = frame_change.phase
-                    elif isinstance(frame_change, pulse.instructions.ShiftPhase):
-                        phase += frame_change.phase
+                phase, frequency = self._calculate_current_frame(frame_changes, phase, frequency)
             frame = PhaseFreqTuple(phase, frequency)
 
-            yield t0, frame, inst
+            yield InstructionTuple(t0, self._dt, frame, inst)
 
-    def get_frame_changes(self) -> Iterator[Tuple[int, PhaseFreqTuple, List[pulse.Instruction]]]:
+    def get_frame_changes(self) -> Iterator[InstructionTuple]:
         """Return frame change type instructions with total frame change amount."""
         sorted_frame_changes = sorted(self._frames.items(), key=lambda x: x[0])
 
-        phase = self.init_phase
-        frequency = self.init_frequency
-        for t0, insts in sorted_frame_changes:
+        phase = self._init_phase
+        frequency = self._init_frequency
+        for t0, frame_changes in sorted_frame_changes:
             pre_phase = phase
             pre_frequency = frequency
-            for inst in insts:
-                if isinstance(inst, pulse.instructions.SetFrequency):
-                    frequency = inst.frequency
-                elif isinstance(inst, pulse.instructions.ShiftFrequency):
-                    frequency += inst.frequency
-                elif isinstance(inst, pulse.instructions.SetPhase):
-                    phase = inst.phase
-                elif isinstance(inst, pulse.instructions.ShiftPhase):
-                    phase += inst.phase
+            phase, frequency = self._calculate_current_frame(frame_changes, phase, frequency)
             frame = PhaseFreqTuple(phase - pre_phase, frequency - pre_frequency)
 
-            yield t0, frame, insts
+            yield InstructionTuple(t0, self._dt, frame, frame_changes)
+
+    @staticmethod
+    def _calculate_current_frame(frame_changes: List[pulse.instructions.Instruction],
```

Comment:
Why is this a static method? It will never need to know about class info?

Reply:
Maybe this is bad function name. Currently we have 4 types of instructions, i.e. shift/set phase/frequency, and this function just manipulates phase of frequency value according to instruction type.

Reply:
If it doesn't need to be static, perhaps we should just make it a class method for consistency.